### DISCUSSION TITLE
About audio in html5 mobile

### DISCUSSION THREAD

**Discussion by moisesjpelaez** [MoritzBrueckner/aura#10] [Q&A] [answered]
I see there's a limitation in [Aura.hx](https://github.com/MoritzBrueckner/aura/blob/master/Sources/aura/Aura.hx) with mobile browsers. Is it possible to do the effects support manually or does it completely depend on Kha?
 
I'm also interested to contribute to this library once I get more experience with Armory and Kha, is there any recommended read where I can learn more about audio implementation? I'd like to understand better this library's structure before contributing with code.

**Comment by MoritzBrueckner** [comment, marked as answer]
Hi, sorry for my late answer. I guess you are referring to the following specific comment?
https://github.com/MoritzBrueckner/aura/blob/a14a1c9ff59743b5f5a67bc18a535a9f58118f2f/Sources/aura/Aura.hx#L73-L78

What I mean with "custom audio callback" in that comment is a custom `kha.audio2.Audio.audioCallback` function that is (indirectly) called by the audio driver to request more audio samples from the application. So whenever the driver needs more samples for the audio output device, it asks Aura to write them into a buffer in the [audio callback](https://github.com/MoritzBrueckner/aura/blob/a14a1c9ff59743b5f5a67bc18a535a9f58118f2f/Sources/aura/Aura.hx#L358).

Unfortunately, the Kha implementation for mobile browsers does not use such a callback, so Aura also can't use it. Instead, the sounds are played via some other browser API (also via a Kha detour) and Aura only regularly updates some properties of the sounds which would otherwise happen in the audio callback (that's what [this line](https://github.com/MoritzBrueckner/aura/blob/a14a1c9ff59743b5f5a67bc18a535a9f58118f2f/Sources/aura/Aura.hx#L77) is for). I believe you can find the Kha implementation for the mobile html5 audio [here](https://github.com/Kode/Kha/blob/main/Backends/HTML5/kha/js/MobileWebAudioChannel.hx). The mobile audio channel is created [here](https://github.com/MoritzBrueckner/aura/blob/a14a1c9ff59743b5f5a67bc18a535a9f58118f2f/Sources/aura/Aura.hx#L318), and it looks like something similar might be missing in `Aura.createUncompBufferChannel()` – it has been a while since I last tested mobile html5 audio so I can't say this for sure...

I can't really tell you the exact reason(s) why Kha does not play audio the same way on mobile browsers as on desktop browsers, I'm not even sure whether I knew it some time ago or whether I never really investigated it. There seems to be a bit of context [here](https://github.com/MoritzBrueckner/aura/blob/a14a1c9ff59743b5f5a67bc18a535a9f58118f2f/Sources/aura/channels/Html5StreamChannel.hx#L144-L148).

Nonetheless, I believe there are some potential solutions for your problem, but they are not easy to implement and they have some big drawbacks:
- Try to use a [`ScriptProcessorNode`](https://developer.mozilla.org/en-US/docs/Web/API/ScriptProcessorNode) to implement the audio callback. I believe this is also used for the non-html5 Kha implementation. Unfortunately, I don't know how well this is supported on mobile browsers, so using this might cause the same problems that Kha worked around by providing mobile-specific audio code... And even worse: using `ScriptProcessorNode`s is deprecated...
- Use [`AudioWorklet`s](https://developer.mozilla.org/en-US/docs/Web/API/AudioWorklet) instead. This is the modern replacement of `ScriptProcessorNode` that works in a separate audio thread, but I also can't tell you how well this is supported on mobile browsers. Sooner or later, Aura should use audio worklets for html5 audio on both mobile and desktop devices, but it would require big changes to Aura (or Kha, if Kha was using it) because the audio worklet code needs to reside in a different JavaScript file to not create a bunch of thread-safety problems. To achieve this, we would need to compile the Haxe code into two files, which could be complicated...

> I'm also interested to contribute to this library once I get more experience with Armory and Kha, is there any recommended read where I can learn more about audio implementation? I'd like to understand better this library's structure before contributing with code.

Very cool! There is almost no documentation about the implementation at the moment (other than some code comments and maybe some hints in the [wiki](https://github.com/MoritzBrueckner/aura/wiki)), but I'm happy to explain it or to write some documentation when I get time. If you have specific questions, feel free to open more discussions here :)

Some general notes about the implementation:
- You should be able to find pretty much all playback-related code by following the audio callback implementation
- Memory allocation in the audio thread code (all code indirectly called by the audio callback) is a no-go (this can be tricky sometimes). On some targets, garbage collection is even turned off in those parts of the code
- Channels have a handle that resides in the main thread (if the application uses an audio thread) and an actual implementation in the audio thread that communicate via messages. This still needs to be implemented for DSP effects, and the message system probably also isn't completely thread-safe at the moment...
- The package structure is pretty chaotic at the moment, especially `aura.types` (which does not make much sense) and `aura.utils`.
